lib/vector-core/src/fanout.rs

```diff
@@ -63,165 +62,330 @@ impl Fanout {
-                // Pending so it's fine to return Pending here since the control
-                // channel will notify current task when it receives a message.
-                None => return Poll::Pending,
-            }
+#[pin_project]
+struct SendGroup {
+    #[pin]
+    sends: Vec<(ComponentKey, SendOp)>,
+    waker: Arc<AtomicWaker>,
+}
+
+impl SendGroup {
+    fn with_capacity(capacity: usize) -> Self {
+        Self {
+            sends: Vec::with_capacity(capacity),
+            waker: Arc::new(AtomicWaker::new()),
         }
+    }
 
-        this.i = 0;
+    fn push(&mut self, id: ComponentKey, sink: BufferSender<Event>, input: Vec<Event>) {
+        let send = SendOp::new(sink, input, Arc::clone(&self.waker));
+        self.sends.push((id, send));
+    }
 
-        Poll::Ready(Ok(()))
+    fn remove(&mut self, id: &ComponentKey) {
+        if let Some(i) = self.sends.iter().position(|(n, _)| n == id) {
+            let (_id, _removed) = self.sends.remove(i);
+        } else {
+            panic!("Tried to remove a sink that's not present");
+        }
     }
 
-    fn start_send(mut self: Pin<&mut Self>, item: Event) -> Result<(), ()> {
-        let mut items = vec![item; self.sinks.len()];
+    fn replace(&mut self, id: &ComponentKey, sink: BufferSender<Event>) {
+        if let Some((_, send)) = self.sends.iter_mut().find(|(n, _)| n == id) {
+            send.replace(sink);
+            // This may have unpaused a send operation, so make sure it is woken up.
+            self.waker.wake();
+        } else {
+            panic!("Tried to replace a sink that's not already present");
+        }
+    }
 
-        let mut i = 1;
-        while let Some((_, sink)) = self.sinks.get_mut(i) {
-            if let Some(sink) = sink.as_mut() {
-                let item = items.pop().unwrap();
-                if sink.as_mut().start_send(item).is_err() {
-                    self.handle_sink_error(i)?;
-                    continue;
-                }
-            }
-            i += 1;
+    fn pause(&mut self, id: &ComponentKey) {
+        if let Some((_, send)) = self.sends.iter_mut().find(|(n, _)| n == id) {
+            send.pause();
+        } else {
+            panic!("Tried to pause a sink that's not already present");
         }
+    }
+}
 
-        if let Some((_, sink)) = self.sinks.first_mut() {
-            if let Some(sink) = sink.as_mut() {
-                let item = items.pop().unwrap();
-                if sink.as_mut().start_send(item).is_err() {
-                    self.handle_sink_error(0)?;
-                }
+impl Future for SendGroup {
+    type Output = Vec<(ComponentKey, Option<BufferSender<Event>>)>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut pending = false;
+        let this = self.as_mut().project();
+        for (_key, send) in this.sends.get_mut() {
+            let send = Pin::new(send);
```

Comment:
Instead of pinning each loop here, would it make sense to pin inside `self.sends` the once?

Reply:
Maybe? I tried a number of things to project a `Pin` into a `Vec`, and this was the simplest I could come up with. It's very possible there's a better way.

Reply:
Actually, thinking about this more and examining the code paths in `Pin`, I don't think it matters overmuch. So, if a notion comes along to push this up I'd say take it, else don't worry about it.